README.md
# Dedicate Redis services on AWS or OpenStack

This is a simple `bosh` CLI plugin to create and delete dedicated redis services and provide a simple URI. It runs on AWS or OpenStack via bosh.

Example create/delete scenario that includes binding the redis service to a Cloud Foundry application via environment variables:

```
$ bosh prepare redis
$ bosh create redis
$ bosh show redis uri
redis://:c1da049a75b3@0.redis.default.demoredis.microbosh:6379/0
$ cf set-env myapp REDIS_URI redis://:c1da049a75b3@0.redis.default.redis-123.microbosh:6379/0

$ cf unset-env myapp REDIS_URI
$ bosh delete redis
```

The resulting redis servers can be discovered and accessed by any systems (such as Cloud Foundry applications) using the same bosh (or microbosh) or a connected DNS.

## Why not use Cloud Foundry v2 services?

See [discussion](https://groups.google.com/a/cloudfoundry.org/d/msg/bosh-users/yoXGZMcWduc/8xocVYxpKakJ) on vcap-dev mailing list. 

## Requirements

You must use the same BOSH for deploying redis as you use to deploy Cloud Foundry (unless your DNS is configured to be shared across the two BOSH).

It is also requried that you have login access to the same BOSH being used to deploy your Cloud Foundry.

Confirm this by running:

```
$ bosh status
$ bosh deployments
```

The former will confirm you are targeting a bosh. The latter will display the deployments. One of which should be your Cloud Foundry.

## Installation

Install via RubyGems:

```
$ gem install bosh_cli "~> 1.5.0.pre" --source https://s3.amazonaws.com/bosh-jenkins-gems/ 
$ gem install bosh_cli_plugin_redis
```

The `bosh_cli` gem is currently only available from S3, rather than RubyGem itself. So it needs to be installed first.

## Usage

Each time you install the latest `redis-cf-plugin` you will want to re-upload the latest available redis release to your bosh. If no newer release is available then nothing good nor bad will occur.

```
$ bosh prepare redis
Uploading new redis release to bosh...
```

To create/provision a new redis service you run the following command. By default, it will select the smallest known instance size.

```
$ bosh create redis
$ bosh create redis --size small
$ bosh create redis --size medium
$ bosh create redis --size large
$ bosh create redis --size xlarge
```

By default the redis server is assigned a 4096 Mb persistent volume/disk. To change this value use `--disk`:

```
$ bosh create redis myapp-redis --disk 8192
```

NOTE: By default, the `default` security group is used. It must have port `6379` open.

To chose a different security group, use the `--security-group` option:

```
$ bosh create redis --security-group redis-server
```

To see the list of available instance sizes or to edit the list of available instance size, see the section "Customizing" below.

* TODO - how to show available instance sizes
* TODO - how to update a redis server to a different instance size/flavor
* TODO - how to update the persistent disk for the redis server

## Customizing

* TODO - how to edit available instance sizes (via the bosh deployment file templates)

## Releasing new plugin gem versions

There are two reasons to release new versions of this plugin.

1. Package the latest [redis-boshrelease](https://github.com/cloudfoundry-community/redis-boshrelease) bosh release (which describes how the redis service is implemented)
2. New features or bug fixes to the plugin

To package the latest "final release" of the redis bosh release into this source repository, run the following command:

```
$ cd /path/to/releases
$ git clone https://github.com/cloudfoundry-community/redis-boshrelease.git
$ cd -
$ rake bosh:release:import[/path/to/releases/redis-boshrelease]
# for zsh shell quotes are required around rake arguments:
$ rake bosh:release:import'[/path/to/releases/redis-boshrelease]'
```

Note: only the latest "final release" will be packaged. See https://github.com/cloudfoundry-community/redis-boshrelease#readme for information on creating new bosh releases.

To locally test the plugin (`bosh` cli loads plugins from its local path automatically):

```
$ cf /path/to/bosh_cli_plugin_redis
$ bosh redis
```

To release a new version of the plugin as a RubyGem:

1. Edit `bosh_cli_plugin_redis.gemspec` to update the major or minor or patch version.
2. Run the release command:

```
$ rake release
```

## Contributing

For fixes or features to the `bosh_cli_plugin_redis` (`bosh redis`) plugin:

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

For fixes or features to the redis bosh release, please visit https://github.com/cloudfoundry-community/redis-boshrelease. Final releases of `redis-boshrelease` will be distributed in future gem versions of this plugin.


ChangeLog.md
# Change Log


## v0.2 - bosh_cli_plugin_redis

Converted into a bosh plugin. Prefix "bosh" to the sub-commands below:

    prepare redis 
        Prepare bosh for deploying one or more Redis services 

    create redis [--name redis-<timestamp>] [--size small] [--disk 4096] 
                 [--security-group default] 
        Create a Redis service deployed upon target bosh 
        --name redis-<timestamp> Unique bosh deployment name 
        --size small             Size of provisioned VMs 
        --disk 4096              Size of persistent disk (Mb) 
        --security-group default Security group to assign to provisioned VMs 

    show redis uri 
        Show the redis URI for connection via bosh DNS 

    delete redis 
        Delete current Redis service 

Note: the `cf bind-redis-env-var` command has been removed as it is a) specific to a Cloud Foundry user not a bosh user; b) kinda crappy "services" integration. Perhaps move it back into a cf plugin in future.

Other features/fixes:

* Validates the resource size with available sizes in the template [v0.2.1]
* `bosh create redis --disk 10000` to disk persistent disk size (Mb) [v0.2.2]
* Fix #1 - pass properties.redis.persistent_disk to templates [v0.2.3]

## v0.1 - redis-cf-plugin

Initial release! The initial commands offered are:

    cf create-redis               Create a Redis service deployed upon target bosh
    cf show-redis-uri             Show the redis URI for connection via bosh DNS
    cf bind-redis-env-var APP     Bind current Redis service URI to current app via env variable
    cf delete-redis               Delete current Redis service

The redis service URI uses bosh DNS. As such it must be deployed into the same bosh being used for Cloud Foundry.


lib/redis-cf-plugin/plugin.rb
require "yaml"

# for the #sh helper
require "rake"
require "rake/file_utils"

module RedisCfPlugin
  class Plugin < CF::CLI
    include FileUtils

    desc "Prepare target bosh for deploying one or more Redis services"
    group :services, :manage
    def prepare_redis
      within_bosh_release do
        # the releases/index.yml contains all the available release versions in an unordered
        # hash of hashes in YAML format:
        #     --- 
        #     builds: 
        #       af61f03c5ad6327e0795402f1c458f2fc6f21201: 
        #         version: 3
        #       39c029d0af9effc6913f3333434b894ff6433638: 
        #         version: 1
        #       5f5d0a7fb577fec3c09408c94f7abbe2d52a042c: 
        #         version: 4
        #       f044d47e0183f084db9dac5a6ef00d7bd21c8451: 
        #         version: 2
        release_index = YAML.load_file("releases/index.yml")
        latest_version = release_index["builds"].values.inject(0) do |max_version, release|
          version = release["version"]
          max_version < version ? version : max_version
        end

        sh "bosh upload release releases/*-#{latest_version}.yml"
      end
    end

    # TODO - size must be valid
    # TODO - name must be unique (cf services & bosh deployments)

    desc "Create a Redis service deployed upon target bosh"
    group :services, :manage
    input :name, desc: "Unique name for service (within bosh & cloud foundry)"
    input :size, desc: "Size of provisioned VMs", default: "small"
    input :security_group, desc: "Security group to assign to provisioned VMs", default: "default"
    def create_redis
      require "cli" # bosh_cli's director client library
      service_name = input[:name] || "redis-#{Time.now.to_i}"
      resource_size = input[:size]
      security_group = input[:security_group]

      bosh_uuid = bosh_director_client.get_status["uuid"]
      bosh_cpi = bosh_director_client.get_status["cpi"]

      template_file = File.join(bosh_release_dir, "templates", bosh_cpi, "single_vm.yml.erb")

      # Create an initial deployment file; upon which the CPI-specific template will be applied below
      # Initial file will look like:
      # ---
      # name: NAME
      # director_uuid: UUID
      # networks: {}
      # properties:
      #   redis:
      #     resource: medium
      #     security_group: redis-server
      deployment_file = "deployments/redis/#{service_name}.yml"
      sh "mkdir -p deployments/redis"
      line "Creating deployment file #{deployment_file}..."
      File.open(deployment_file, "w") do |file|
        file << {
          "name" => service_name,
          "director_uuid" => bosh_uuid,
          "networks" => {},
          "properties" => {
            "redis" => {
              "resource" => resource_size,
              "security_group" => security_group
            }
          }
        }.to_yaml
      end

      sh "bosh deployment #{deployment_file}"
      sh "bosh -n diff #{template_file}"
      sh "bosh -n deploy"
    end

    desc "Show the redis URI for connection via bosh DNS"
    group :services, :manage
    def show_redis_uri
      load_bosh_and_validate_current_deployment
      print service_uri
    end

    desc "Bind current Redis service URI to current app via env variable"
    input :app, argument: "required", desc: "Application to immediately bind to", from_given: by_name(:app)
    input :env_var, desc: "Environment variable to bind redis URI (default $REDIS_URI)", default: "REDIS_URI"
    group :services, :manage
    def bind_redis_env_var
      load_bosh_and_validate_current_deployment
      env_var = input[:env_var]
      app = input[:app]
      invoke :set_env, app: app, name: env_var, value: service_uri, restart: true
    end

    desc "Delete current Redis service"
    group :services, :manage
    def delete_redis
      load_bosh_and_validate_current_deployment
      sh "bosh delete deployment #{deployment_name}"
    end

    protected
    def release_name
      "redis"
    end

    def bosh_release_dir
      File.expand_path("../../../bosh_release", __FILE__)
    end

    def within_bosh_release(&block)
      chdir(bosh_release_dir, &block)
    end

    def bosh_director_client
      @bosh_director_client ||= begin
        command = ::Bosh::Cli::Command::Deployment.new
        command.send(:auth_required)
        command.director
      end
    end

    def deployment_file
      @current_deployment_file ||= ::Bosh::Cli::Command::Deployment.new.deployment
    end

    def load_bosh_and_validate_current_deployment
      require "cli" # bosh_cli's director client library
      unless File.exists?(deployment_file)
        fail "Target deployment file no longer exists: #{deployment_file}"
      end
      @deployment = YAML.load_file(deployment_file)
      unless @deployment["release"] && @deployment["release"]["name"] == release_name
        fail "Target deployment file is not for redis service: #{deployment_file}"
      end
    end

    def deployment_name
      @deployment["name"]
    end

    def service_uri
      password = @deployment["properties"]["redis"]["password"]
      port = @deployment["properties"]["redis"]["port"]
      db = 0
      "redis://:#{password}@#{server_host}:#{port}/#{db}"
    end

    # returns the first DNS entry for the running instance
    def server_host
      @server_host ||= begin
        vms = bosh_director_client.fetch_vm_state(deployment_name, use_cache: false)
        if vms.empty?
          fail "Deployment has no running instances"
        end
        if vms.size > 1
          warn "Deployment has more than 1 running instance (#{vms.size}); using first instance"
        end
        vm = vms.first
        vm["dns"].first
      end
    end
  end
end


lib/bosh/cli/commands/redis.rb
require "yaml"

# for the #sh helper
require "rake"
require "rake/file_utils"

module Bosh::Cli::Command
  class Redis < Base
    include FileUtils
    include Bosh::Cli::Validation

    usage "redis"
    desc  "show micro bosh sub-commands"
    def redis_help
      say("bosh redis sub-commands:")
      nl
      cmds = Bosh::Cli::Config.commands.values.find_all {|c|
        c.usage =~ /redis/
      }
      Bosh::Cli::Command::Help.list_commands(cmds)
    end

    usage "prepare redis"
    desc "Prepare bosh for deploying one or more Redis services"
    def prepare_redis
      within_bosh_release do
        # the releases/index.yml contains all the available release versions in an unordered
        # hash of hashes in YAML format:
        #     --- 
        #     builds: 
        #       af61f03c5ad6327e0795402f1c458f2fc6f21201: 
        #         version: 3
        #       39c029d0af9effc6913f3333434b894ff6433638: 
        #         version: 1
        #       5f5d0a7fb577fec3c09408c94f7abbe2d52a042c: 
        #         version: 4
        #       f044d47e0183f084db9dac5a6ef00d7bd21c8451: 
        #         version: 2
        release_index = YAML.load_file("releases/index.yml")
        latest_version = release_index["builds"].values.inject(0) do |max_version, release|
          version = release["version"]
          max_version < version ? version : max_version
        end

        sh "bosh upload release releases/*-#{latest_version}.yml"
      end
    end

    # TODO - size must be valid
    # TODO - name must be unique (cf services & bosh deployments)

    usage "create redis"
    desc "Create a Redis service deployed upon target bosh"
    option "--name redis-<timestamp>", String, "Unique bosh deployment name"
    option "--size small", String, "Size of provisioned VMs"
    option "--disk 4096", Integer, "Size of persistent disk (Mb)"
    option "--security-group default", String, "Security group to assign to provisioned VMs"
    def create_redis
      auth_required

      service_name = options[:name] || default_name
      resource_size = options[:size] || default_size
      persistent_disk = options[:disk] || default_persistent_disk
      security_group = options[:security_group] || default_security_group
      redis_port = 6379

      bosh_status # preload
      nl
      say "CPI: #{bosh_cpi.make_green}"
      say "Deployment name: #{service_name.make_green}"
      say "Resource size: #{validated_resource_size_colored(resource_size)}"
      say "Persistent disk: #{persistent_disk.to_s.make_green}"
      say "Security group: #{security_group.make_green}"
      nl

      step("Validating resource size", "Resource size must be in #{available_resource_sizes.join(', ')}", :non_fatal) do
        available_resource_sizes.include?(resource_size)
      end

      unless confirmed?("Security group exists with ports 22 & #{redis_port}")
        cancel_deployment
      end
      unless confirmed?("Creating redis service")
        cancel_deployment
      end

      raise Bosh::Cli::ValidationHalted unless errors.empty?

      template_file = template_file("single_vm.yml.erb")

      # Create an initial deployment file; upon which the CPI-specific template will be applied below
      # Initial file will look like:
      # ---
      # name: NAME
      # director_uuid: UUID
      # networks: {}
      # properties:
      #   redis:
      #     resource: medium
      #     security_group: redis-server
      deployment_file = "deployments/redis/#{service_name}.yml"
      step("Checking/creating #{redis_deployments_store_path} for deployment files",
           "Failed to create #{redis_deployments_store_path} for deployment files", :fatal) do
        mkdir_p(redis_deployments_store_path)
      end

      step("Creating deployment file #{deployment_file}",
           "Failed to create deployment file #{deployment_file}", :fatal) do
        File.open(deployment_file, "w") do |file|
          file << {
            "name" => service_name,
            "director_uuid" => bosh_uuid,
            "networks" => {},
            "properties" => {
              "redis" => {
                "resource" => resource_size,
                "security_group" => security_group,
                "persistent_disk" => persistent_disk
              }
            }
          }.to_yaml
        end

        stdout = Bosh::Cli::Config.output
        Bosh::Cli::Config.output = nil
        deployment_cmd(non_interactive: true).set_current(deployment_file)
        biff_cmd(non_interactive: true).biff(template_file)
        Bosh::Cli::Config.output = stdout
      end
      # re-set current deployment to show output
      deployment_cmd.set_current(deployment_file)
      deployment_cmd(non_interactive: options[:non_interactive]).perform
    rescue Bosh::Cli::ValidationHalted
      errors.each do |error|
        say error.make_red
      end
    end

    usage "show redis uri"
    desc "Show the redis URI for connection via bosh DNS"
    def show_redis_uri
      load_bosh_and_validate_current_deployment
      print service_uri
    end

    usage "delete redis"
    desc "Delete current Redis service"
    def delete_redis
      load_bosh_and_validate_current_deployment
      deployment_cmd(non_interactive: options[:non_interactive]).delete(deployment_name)
    end

    protected

    def default_name
      "redis-#{Time.now.to_i}"
    end

    def default_size
      "small"
    end

    def default_persistent_disk
      4096
    end

    def default_security_group
      "default"
    end

    def release_name
      "redis"
    end

    def redis_deployments_store_path
      "deployments/redis"
    end

    def bosh_release_dir
      File.expand_path("../../../../../bosh_release", __FILE__)
    end

    def within_bosh_release(&block)
      chdir(bosh_release_dir, &block)
    end

    def template_file(file)
      File.join(bosh_release_dir, "templates", bosh_cpi, file)
    end

    def bosh_release_spec
      @bosh_release_spec ||= begin
        unless File.exists?(template_file("spec"))
          err "Bosh release templates missing 'spec'"
        end
        YAML.load_file(template_file("spec"))
      end
    end

    def available_resource_sizes
      resources = bosh_release_spec["resources"]
      if resources && resources.is_a?(Array) && resources.first.is_a?(String)
        resources
      else
        err "template spec needs 'resources' key with list of resource pool names available"
      end
    end

    # If resource_size is within +available_resource_sizes+ then display it in green;
    # else display it in red.
    def validated_resource_size_colored(resource_size)
      available_resource_sizes.include?(resource_size) ?
        resource_size.make_green : resource_size.make_red
    end

    def bosh_status
      @bosh_status ||= begin
        step("Fetching bosh information", "Cannot fetch bosh information", :fatal) do
           @bosh_status = bosh_director_client.get_status
        end
        @bosh_status
      end
    end

    def bosh_uuid
      bosh_status["uuid"]
    end

    def bosh_cpi
      bosh_status["cpi"]
    end

    # TODO this is now a bosh cli command itself; use #director
    def bosh_director_client
      director
    end

    def deployment_cmd(options = {})
      cmd = Bosh::Cli::Command::Deployment.new
      options.each do |key, value|
        cmd.add_option key.to_sym, value
      end
      cmd
    end

    def biff_cmd(options = {})
      cmd = Bosh::Cli::Command::Biff.new
      options.each do |key, value|
        cmd.add_option key.to_sym, value
      end
      cmd
    end

    def deployment_file
      deployment
    end

    # TODO use bosh cli helpers to validate/require this
    def load_bosh_and_validate_current_deployment
      auth_required
      unless File.exists?(deployment_file)
        err "Target deployment file no longer exists: #{deployment_file}"
      end
      @deployment = YAML.load_file(deployment_file)
      unless @deployment["release"] && @deployment["release"]["name"] == release_name
        err "Target deployment file is not for redis service: #{deployment_file}"
      end
    end

    def deployment_name
      @deployment["name"]
    end

    def service_uri
      password = @deployment["properties"]["redis"]["password"]
      port = @deployment["properties"]["redis"]["port"]
      db = 0
      "redis://:#{password}@#{server_host}:#{port}/#{db}"
    end

    # returns the first DNS entry for the running instance
    def server_host
      @server_host ||= begin
        vms = bosh_director_client.fetch_vm_state(deployment_name, use_cache: false)
        if vms.empty?
          err "Deployment has no running instances"
        end
        if vms.size > 1
          say "#{"WARNING!".make_red} Deployment has more than 1 running instance (#{vms.size}); using first instance"
        end
        vm = vms.first
        # bug in bosh dns means old dns entries might still point to this new VM
        vm["dns"].select { |dns| dns =~ %r{#{deployment_name}}  }.first
      end
    end
  end
end
